stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,573 @@
+			}
+			defer streams[i].fifo.Close()
+			streams[i].listener, err = net.Listen(ep.Domain, ep.Address)
+			if err != nil {
+				return fmt.Errorf("unable to open %s listener: %s", entry.streamName, err)
+			}
+			defer streams[i].listener.Close()
+			endpoints.Targets = append(endpoints.Targets, ep)
+			diag.Printf("Listening for %s on %s\n", entry.streamName, ep.Address)
+		}
+	}
+
+	// write available endpoints to status file
+	sf, err := os.OpenFile(statusFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	err = json.NewEncoder(sf).Encode(endpoints)
+	_ = sf.Sync()
+	_ = sf.Close()
+	if err != nil {
+		return err
+	}
+
+	c := make(chan error)
+
+	// TODO(lucab): finalize custom logging modes
+	logMode := os.Getenv("STAGE1_LOGMODE")
+	var logFile *os.File
+	switch logMode {
+	case "k8s-plain":
+		var err error
+		// TODO(lucab): check what should be the target path with Euan
+		logTarget := filepath.Join(pathPrefix, appName, "logfile")
+		logFile, err = os.OpenFile(logTarget, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		defer logFile.Close()
+	}
+
+	// proxy stdin
+	if streams[0].fifo != nil && streams[0].listener != nil {
+		clients := make(chan net.Conn)
+		go acceptConn(streams[0].listener, clients, "stdin")
+		go muxInput(clients, streams[0].fifo)
+	}
+
+	// proxy stdout
+	if streams[1].fifo != nil && streams[1].listener != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(streams[1].fifo, lines, c)
+		go acceptConn(streams[1].listener, clients, "stdout")
+		go muxOutput("stdout", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	// proxy stderr
+	if streams[2].fifo != nil && streams[2].listener != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(streams[2].fifo, lines, c)
+		go acceptConn(streams[2].listener, clients, "stderr")
+		go muxOutput("stderr", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	// If nothing else fails, iomux service will be waiting here forever
+	// and be terminated by systemd only when the main application exits.
+	return <-c
+}
+
+// bufferLine buffers and queues a single line from a Reader to a multiplexer
+// If reading from src fails, it hard-fails and propagates the error back.
+func bufferLine(src io.Reader, c chan<- []byte, ec chan<- error) {
+	rd := bufio.NewReader(src)
+	for {
+		lineOut, err := rd.ReadBytes('\n')
+		if len(lineOut) == 0 && err != nil {
+			ec <- err
+		}
+		c <- lineOut
+	}
+}
+
+// acceptConn accepts a single client and queues it for further proxying
+// It is never canceled explicitly, as it is bound to the lifetime of the main process.
+func acceptConn(socket net.Listener, c chan<- net.Conn, stream string) {
+	for {
+		conn, err := socket.Accept()
```

Comment:
As far as I see we never close this socket. I guess when the process quits, the socket will be closed implicitly, but in case of an error implying process exit we could close it.